ksql-engine/src/main/java/io/confluent/ksql/codegen/FunctionArguments.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.codegen;
+
+import org.apache.kafka.connect.data.Schema;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+public class FunctionArguments {
+  private Stack<List<Schema.Type>> args = new Stack<>();
+
+  public void beginFunction() {
+    args.push(new ArrayList<>());
+  }
+
+  public void addArgumentType(final Schema.Type type) {
+    if (!args.isEmpty()) {
+      args.peek().add(type);
+    }
+  }
+
+  public List<Schema.Type> endFunction() {
+    return args.pop();
+  }
+
+  public void mergeArguments(final int from) {
+    final List<Schema.Type> functionArgs = args.peek();
+    final Schema.Type first = functionArgs.remove(from);
+    final Schema.Type second = functionArgs.remove(from);
+    functionArgs.add(first.ordinal() < second.ordinal() ? second : first);
```

Comment:
It looks like what we're doing here is using the enumeration order of Type to resolve the resulting type of arithmetic binary expressions. Few issues here:
   - Using the ordinality feels brittle to me. Is that order guaranteed never to change, or to at least always resolve the type for any arithmetic expression correctly?
   - Does that order make sense for any operator? I think its ok for the ones we have now, but maybe not for other operators we may want to add (e.g. bit-shift). It seems like an easy thing to miss if the schema resolution is so decoupled from the code that sees the actual expressions. I think it would be better to pass in the resolved schema.